Feature/configs for latest tslint versions

christophercr opened the pull request:
ISSUES CLOSED: #15

## PR Checklist

Please check if your PR fulfills the following requirements:

-   [X] The commit message follows our guidelines: https://github.com/NationalBankBelgium/code-style/blob/master/CONTRIBUTING.md#-commit-message-guidelines
-   [ ] Tests for the changes have been added (for bug fixes / features)
-   [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[X] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #15 

## What is the new behavior?

code-style now contains the configs for TSLint 5.11 to 6.1

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

christophercr commented on lib/tslint/6.0.x/tslint.json:
```diff
@@ -0,0 +1,225 @@
+    "no-async-without-await": true,
+    "invalid-void": true,
+    // TSLint rules that were previously on tslint:recommended prior to v6.0
+    //"arrow-parens": true,
```
@SuperITMan @nicanac I'm not sure what to do with these rules. They were removed from `tslint:recommended` in TSLint v6.0 in https://github.com/palantir/tslint/pull/4871/files and https://github.com/palantir/tslint/pull/4312/files

Normally that shouldn't affect but, in fact in Stark we extend our configs from `tslint:latest` which actually extends from `tslint:recommended` (see https://github.com/palantir/tslint/blob/master/src/configs/latest.ts) therefore I was thinking on adding them here... but we should also consider that these rules are disabled by `tslint-config-prettier` (which we use in Stark :p).

So.. what do you think we should do here? Should we add the rules here anyway to ensure that any project using `code-style` without `Prettier` keeps the same rules? or we just don't care so we don't add them here so any impacted project would need to add those by themselves?

christophercr replied:
I've created PR https://github.com/NationalBankBelgium/stark/pull/1655 where you can see all the adaptations needed in Stark based on the new TSLint rules/options enabled from 5.10 to 6.0.

christophercr replied:
The rules listed in here are the ones that were removed from the `tslint:recommended` and in fact those rules are disabled by `tslint-config-prettier` in Stark, hence my question: 
> So.. what do you think we should do here? Should we add the rules here anyway to ensure that any project using code-style without Prettier keeps the same rules? or we just don't care so we don't add them here so any impacted project would need to add those by themselves?

About the `whitespace` rule, it is disabled in our configs anyway ;)

SuperITMan replied:
I would say "we don't care". 
I think we can keep those rules in comments as you did, like this it would be easier for the dev to find the previous rules. 

Maybe we could add an explanation in the README of tsconfig about the fact that we recommend to use `tslint-config-prettier` and `prettier` in addition to the config we provide 😊 

What do you think?

SuperITMan commented:
@christophercr 
I see that you added the following rules in tslint 5.12.x: 
  - `"ban-ts-ignore": true`
  - `"function-constructor": true`
  - `"unnecessary-constructor": true`
  - `"unnecessary-bind": true`

And then you removed the following in tslint 5.14.x:
  - `"no-tautology-expression": true`
  - `"static-this": true`
  - `"unnecessary-else: true`

Could you please add some comments about those changes in your commit messages ? 😊

christophercr commented:
@SuperITMan @nicanac I've updated the commit messages for every version of TSLint ;)

SuperITMan commented on lib/tslint/5.15.x/tslint.json:
```diff
@@ -0,0 +1,191 @@
+    "no-unbound-method": [true, { "ignore-static": true, "whitelist": ["expect"], "allow-typeof": true, "allow-delete": true }],
+    "no-unnecessary-type-assertion": true,
+    "restrict-plus-operands": true,
+    "unnecessary-bind": true
```
Could you maybe add the rule `"no-null-undefined-union": true` ? 
I enabled it in Stark and it does not break anything. 
According to the [documentation](https://palantir.github.io/tslint/rules/no-null-undefined-union/), it's interesting to enable this because it allows the `triple-equals` rule to exist without exceptions.

christophercr replied:
True, I'll add it ;)

christophercr commented on lib/tslint/5.12.x/tslint.json:
```diff
@@ -0,0 +1,188 @@
+    "no-duplicate-switch-case": true,
+    "newline-before-return": false,
+    "newline-per-chained-call": false,
+    "file-name-casing": [true, { "public_api.ts": "snake-case", ".*": "kebab-case" }],
```
In fact I didn't add it to TSLint 5.11.x because the `snake-case` option of that rule was only added until TSLint 5.12, and if we would enable this rule already in TSLint 5.11 without the `snake-case` option then we would get linting errors due to the `public_api.ts` files that we have in Stark.

Do you think we should enable it anyway in TSLint 5.11 anyway?

SuperITMan replied:
You're right! Let's keep it in 5.12 👍

christophercr commented on lib/tslint/5.19.x/tslint.json:
```diff
@@ -0,0 +1,195 @@
+    "unnecessary-else": [true, { "allow-else-if": true }],
+    "no-async-without-await": true,
+    "invalid-void": true,
+
```
Mmm I'm not sure of this. as far as I remember the `for...in` was still needed in some cases where you might have an object that has also some props defined via its prototype. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#Difference_between_for...of_and_for...in.

Moreover, if you check the live demo of `for...of` in the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) and you adapt it to loop over an object like this:

```javascript
const someObject = {first:'a', second:'b', third:'c'};

for (const element of someObject) {
  console.log(element);
}
```

Then you get an error: `Error: someObject is not iterable` 😕 

Whereas if you change the loop for a `for...in` then it works!

SuperITMan replied:
Okok ! Let's keep it as this so ! 😊

christophercr commented:
@SuperITMan @nicanac I've removed the `require-const-for-all-caps` of `variable-name` rule in configs for TSLint lower than 5.14. I've also added the `no-null-undefined-union` rule in configs  of TSLint 5.15 and higher.

I've adapted the commit messages accordingly 👍

christophercr commented:
@nicanac @SuperITMan I've updated this PR by adding the config for the latest version of TSLint released some days ago: `6.1.x` (which is practically the same as for 6.0.x).

Can you please have a look?

christophercr commented on lib/tslint/5.11.x/tslint.json:
```diff
@@ -0,0 +1,177 @@
+    "no-duplicate-switch-case": true,
+    "newline-before-return": false,
+    "newline-per-chained-call": false,
+    "file-name-casing": [false, { ".*": "kebab-case" }],
```
Done ;)

christophercr commented on lib/prettier/README.md:
````diff
@@ -17,3 +17,23 @@ Create a `.prettierrc.js` file and copy the following content inside:
 module.exports = require("@nationalbankbelgium/code-style/prettier/1.16.x");
 ```
+
+## Recommended Additional Configs
````
@SuperITMan I've added this section in Prettier readme as you suggested 👍

christophercr commented on lib/tslint/README.md:
````diff
@@ -10,16 +10,50 @@ The following versions are available (based on **TSLint** version):
 }
 ```
+
+## Recommended Additional Configs
````
@SuperITMan I've added this section in Prettier readme as you suggested 👍

christophercr commented:
@SuperITMan I've adapted this PR with your remarks. Can you have a look?

christophercr commented:
And rebased 😉 

Ready to merge 🎉
